### ISSUE TITLE
Diagnostic for closed files in TSC

### ISSUE THREAD

**Issue by timsofteng** [jose-elias-alvarez/null-ls.nvim#643]
### FAQ

- [X] I have checked the [FAQ](https://github.com/jose-elias-alvarez/null-ls.nvim#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Neovim Version

0.6.1

### Operating System

Void linux

### Minimal config

require("null-ls").setup({
  sources = {
    require("null-ls").builtins.diagnostics.tsc,
    require("null-ls").builtins.formatting.prettier,
  },
  debug=true
})

### Steps to reproduce

1. Clone test repo:
https://github.com/timsofteng/typescript-lsp-server-bug-test

In `src/pages/XxxHomePage` `BlaChild` component has required types. `BlaParent` doesn't send it to child.

2. Open nvim just with only `BlaChild` component or any other component but not `BlaParent`. You can't see error in `BlaParent` until you open this file.

### Expected behavior

TSC should show errors in closed files.

### Actual behavior

TSC show only errors in files which are opened in nvim buffers.

### Debug log

[TRACE Fri 04 Feb 2022 11:06:22 PM EET] ...te/pack/packer/start/null-ls.nvim/lua/null-ls/client.lua:92: starting null-ls client
[TRACE Fri 04 Feb 2022 11:06:22 PM EET] .../site/pack/packer/start/null-ls.nvim/lua/null-ls/rpc.lua:118: received LSP request for method initialize
[DEBUG Fri 04 Feb 2022 11:06:23 PM EET] ...te/pack/packer/start/null-ls.nvim/lua/null-ls/client.lua:147: unable to notify client for method textDocument/didOpen (client not active): {
  textDocument = {
    uri = "file:///home/tim/typescript-lsp-server-bug-test/src/pages/XxxHomePage/BlaChild.tsx"
  }
}
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] .../site/pack/packer/start/null-ls.nvim/lua/null-ls/rpc.lua:143: received LSP notification for method initialized
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] .../site/pack/packer/start/null-ls.nvim/lua/null-ls/rpc.lua:143: received LSP notification for method textDocument/didOpen
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] ...ack/packer/start/null-ls.nvim/lua/null-ls/generators.lua:21: running generators for method NULL_LS_DIAGNOSTICS_ON_OPEN
[DEBUG Fri 04 Feb 2022 11:06:23 PM EET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:335: spawning command [tsc] at /home/tim/typescript-lsp-server-bug-test with args { "--pretty", "false", "--noEmit" }
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:224: error output: nil
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:225: output: 
[41m                                                                               [0m
[41m[37m                This is not the tsc command you are looking for                [0m
[41m                                                                               [0m

To get access to the TypeScript compiler, [34mtsc[0m, from the command line either:

- Use [1mnpm install typescript[0m to first add TypeScript to your project [1mbefore[0m using npx
- Use [1myarn[0m to avoid accidentally running code from un-installed packages

[TRACE Fri 04 Feb 2022 11:06:23 PM EET] ...ck/packer/start/null-ls.nvim/lua/null-ls/diagnostics.lua:156: received diagnostics from source 1
[TRACE Fri 04 Feb 2022 11:06:23 PM EET] ...ck/packer/start/null-ls.nvim/lua/null-ls/diagnostics.lua:157: {}
[TRACE Fri 04 Feb 2022 11:06:27 PM EET] .../site/pack/packer/start/null-ls.nvim/lua/null-ls/rpc.lua:143: received LSP notification for method textDocument/didClose

### Help

No

### Requirements

- [X] I have read and followed the instructions above and understand that my issue will be closed if I did not provide the required information.

**Comment by jose-elias-alvarez** [comment]
What version of `tsc` are you using? The output indicates that it doesn't support the arguments it's receiving, so you may be using an older version. On my end, version 4.5.5 works as expected, both via null-ls and when running `tsc --pretty false --noEmit` from the command line.

**Comment by timsofteng** [comment]
> What version of `tsc` are you using? The output indicates that it doesn't support the arguments it's receiving, so you may be using an older version. On my end, version 4.5.5 works as expected, both via null-ls and when running `tsc --pretty false --noEmit` from the command line.

Thanks!
I've installed `typescript` package globally and looks like now it works with `:lua vim.diagnostic.setqflist()`.
Thanks for your help and politeness.
Just few questions.
How can I trigger it on save and show the results in `Trouble` or status line or whatever?
Can you suggest please some best practice how to trigger it in right way?

**Comment by timsofteng** [comment]
And tsc is so hart in terms of resource usage...
![image](https://user-images.githubusercontent.com/51912173/152655045-b71296b5-deeb-4a4d-8701-f876f8b97660.png)
Sometimes it take whole my processor...

**Comment by timsofteng** [comment]
![image](https://user-images.githubusercontent.com/51912173/152655156-8b96e1e8-af78-4712-a441-0417df1f7aff.png)
Sometimes there are many tsc processes in for one project...

**Comment by jose-elias-alvarez** [comment]
Right, the source will run `tsc` every time you save the file or open a new file, and there's currently no protections against running multiple copies of the same process (see #596 for more details). Either way, that's more about improving null-ls behavior in general, so it looks like it's safe to close this issue.
